File: common/src/main/java/net/shiroha233/roadweaver/features/path/decoration/types/FenceWaypointDecoration.java
package net.shiroha233.roadweaver.features.path.decoration.types;

import net.minecraft.core.BlockPos;
import net.minecraft.world.level.WorldGenLevel;
import net.minecraft.world.level.block.Blocks;
import net.shiroha233.roadweaver.core.model.WoodAssets;
import net.shiroha233.roadweaver.features.path.decoration.base.Decoration;
import net.shiroha233.roadweaver.features.path.decoration.material.BiomeWoodAware;

/**
 * 围栏路标装饰
 */
public class FenceWaypointDecoration extends Decoration implements BiomeWoodAware {
    private WoodAssets wood;

    public FenceWaypointDecoration(BlockPos placePos, WorldGenLevel world) {
        super(placePos, world);
    }

    @Override
    public void place() {
        if (!placeAllowed()) return;
        BlockPos surfacePos = this.getPos();
        WorldGenLevel world = this.getWorld();
        world.setBlock(surfacePos, wood.fence().defaultBlockState(), 3);
        world.setBlock(surfacePos.above(), Blocks.TORCH.defaultBlockState(), 3);
    }

    @Override
    public void setWoodType(WoodAssets assets) {
        this.wood = assets;
    }
}

File: common/src/main/kotlin/net/shiroha233/roadweaver/features/path/decoration/types/FenceWaypointDecoration.kt
package net.shiroha233.roadweaver.features.path.decoration.types

import net.minecraft.core.BlockPos
import net.minecraft.world.level.WorldGenLevel
import net.minecraft.world.level.block.Blocks
import net.shiroha233.roadweaver.features.path.decoration.base.Decoration
import net.shiroha233.roadweaver.features.path.decoration.material.wood.BiomeWoodAware
import net.shiroha233.roadweaver.helpers.Records

class FenceWaypointDecoration(placePos: BlockPos, world: WorldGenLevel) : Decoration(placePos, world), BiomeWoodAware {
    private lateinit var wood: Records.WoodAssets

    override fun place() {
        if (!placeAllowed()) return
        val surfacePos = getPos()
        val world = getWorld()
        world.setBlock(surfacePos, wood.fence.defaultBlockState(), 3)
        world.setBlock(surfacePos.above(), Blocks.TORCH.defaultBlockState(), 3)
    }

    override fun setWoodType(assets: Records.WoodAssets) {
        wood = assets
    }
}
